ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java

```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright 2019 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.cloud.azure;
+
+import com.github.ambry.cloud.CloudDestination;
+import com.github.ambry.cloud.CloudStorageException;
+import com.github.ambry.commons.BlobId;
+import com.microsoft.azure.storage.CloudStorageAccount;
+import com.microsoft.azure.storage.OperationContext;
+import com.microsoft.azure.storage.StorageException;
+import com.microsoft.azure.storage.blob.BlobContainerPublicAccessType;
+import com.microsoft.azure.storage.blob.BlobRequestOptions;
+import com.microsoft.azure.storage.blob.CloudBlobClient;
+import com.microsoft.azure.storage.blob.CloudBlobContainer;
+import com.microsoft.azure.storage.blob.CloudBlockBlob;
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import java.util.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Implementation of {@link CloudDestination} that interacts with Azure Blob Storage service.
+ */
+class AzureCloudDestination implements CloudDestination {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AzureCloudDestination.class);
+  private final CloudStorageAccount azureAccount;
+  private final CloudBlobClient azureBlobClient;
+
+  /**
+   * Construct an Azure cloud destination from a container's replication config.
+   * @param configSpec the config spec to use.
+   * @throws CloudStorageException if the destination could not be created.
+   */
+  AzureCloudDestination(String configSpec) throws CloudStorageException {
+    this(configSpec, null);
+  }
+
+  /**
+   * Construct an Azure cloud destination from a container's replication config and a {@link CloudStorageAccount} instance.
+   * @param configSpec the config spec to use.
+   * @param azureAccount the {@link CloudStorageAccount} to use.
+   * @throws CloudStorageException if the destination could not be created.
+   */
+  AzureCloudDestination(String configSpec, CloudStorageAccount azureAccount) throws CloudStorageException {
```

Comment:
minor: it looks like this can take just `CloudStorageAccount azureAccount` if you change the other constructor to
`this(CloudStorageAccount.parse(configSpec))`?

Reply:
Yes I could do that.  But I prefer to have all the exception-throwing logic and handling in one constructor.

Reply:
Made the change.